### PR TITLE
Fire Mode: Launch NTP in different modes

### PR DESCRIPTION
**Pull request by 0nko**
Task/Issue URL: https://app.asana.com/1/137249556945/project/1207418217763355/task/1215275434974447?focus=true

### Description

This PR implements the functionality that allows launching NTPs specifically in Regular or Browser mode.

### Steps to test this PR

- [ ] Enable Fire tabs FF in dev settings
- [ ] Make Search & Duck.ai option is enabled in AI settings
- [ ] Open a new Duck.ai chat tab in Regular mode
- [ ] Tap on the + button and tap on "New Fire tab"
- [ ] Verify a new tab is launched and the browser is now in Fire Mode (Fire mode theme is used)
- [ ] While in Fire Mode, open Duck.ai chat tab
- [ ] Tap on the + button and tap on "New tab"
- [ ] Verify a new tab is launched and the browser is now in Regular Mode (regular theme is used)

---

> [!NOTE]
> **Medium Risk**
> Touches core BrowserActivity lifecycle, mode switching, and intent handling; incorrect deferral or replay could drop tabs or mishandle launch intents, though ProcessIntent consumption is explicitly guarded.
> 
> **Overview**
> Adds **mode-aware new tab launches** from Duck.ai’s “+” menu: **New tab** switches to Regular mode first; **New Fire tab** switches to Fire mode first.
> 
> Replaces the old FIRE→REGULAR-only deferred `Intent` stash with a general **`switchModeThen`** flow: pending work is stored as **`PendingModeSwitch`** (`ProcessIntent` or `OpenNewTab`), saved in instance state across activity recreate, then replayed in `showWebContent`. **`launchNewTab`** gains an optional **`browserMode`** parameter (default: current mode) and always goes through that deferral path when a switch is needed.
> 
> New **`PendingModeSwitch.kt`** defines bundle round-tripping; **`PendingModeSwitchTest`** covers encode/decode edge cases. **`runAction`** documents that carried intents are consumed via `processIntent`, not re-launched.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 407bc24427f1adfed325193d32ae894d324faa97. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by 0nko** [comment]
* **#9078** <a href="https://app.graphite.com/github/pr/duckduckgo/Android/9078?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#9053** <a href="https://app.graphite.com/github/pr/duckduckgo/Android/9053?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#9021** <a href="https://app.graphite.com/github/pr/duckduckgo/Android/9021?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#9001** <a href="https://app.graphite.com/github/pr/duckduckgo/Android/9001?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8917** <a href="https://app.graphite.com/github/pr/duckduckgo/Android/8917?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8902** <a href="https://app.graphite.com/github/pr/duckduckgo/Android/8902?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8792** <a href="https://app.graphite.com/github/pr/duckduckgo/Android/8792?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8791** <a href="https://app.graphite.com/github/pr/duckduckgo/Android/8791?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8748** <a href="https://app.graphite.com/github/pr/duckduckgo/Android/8748?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8742** <a href="https://app.graphite.com/github/pr/duckduckgo/Android/8742?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/duckduckgo/Android/8742?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#8725** <a href="https://app.graphite.com/github/pr/duckduckgo/Android/8725?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8714** <a href="https://app.graphite.com/github/pr/duckduckgo/Android/8714?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `develop`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: app/src/main/java/com/duckduckgo/app/browser/PendingModeSwitch.kt [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            bundle.putString(KEY_SOURCE_TAB_ID, pendingAction.sourceTabId)
+            bundle.putBoolean(KEY_SKIP_HOME, pendingAction.skipHome)
+            bundle.putBoolean(KEY_IS_EXTERNAL, pendingAction.isExternal)
+        }
+    }
+    return bundle
+}
+
+internal fun Bundle.toPendingModeSwitch(): PendingModeSwitch? {
+    val targetMode = getString(KEY_TARGET_MODE)?.let { BrowserMode.valueOf(it) } ?: return null
```
[the remark this answers is not in the document]
Comment (0nko) [reply app/src/main/java/com/duckduckgo/app/browser/PendingModeSwitch.kt]: Fixed in 7317f7321b. `toPendingModeSwitch()` is a fail-soft parser — every other malformed-bundle branch returns `null` (missing intent, unknown action) and a null result is handled as "no pending action". The bare `valueOf` was the only branch that threw, so it now parses with `runCatching { BrowserMode.valueOf(it) }.getOrNull()` and degrades to null like the rest instead of crashing on restore.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserActivity.kt [hunk 2]
```diff
@@ -1658,6 +1654,57 @@ open class BrowserActivity : DuckDuckGoActivity() {
...
+     * [pendingModeSwitch], carried across the recreate via [onSaveInstanceState], and replayed by
+     * [BrowserStateRenderer.showWebContent].
+     *
+     * The replay calls this same method, so it is self-correcting: on the recreated instance the
+     * mode matches and the action runs; were it ever to run before the recreate, it simply re-stashes.
+     */
+    private fun switchModeThen(targetMode: BrowserMode, action: PendingAction) {
+        if (targetMode == currentBrowserMode) {
+            runAction(action)
+            return
```
[the remark this answers is not in the document]
Comment (0nko) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserActivity.kt:1720]: This is a false positive — `currentBrowserMode` and the repository `tabManager` writes to cannot diverge within an instance.

Both resolve from the same `@SingleInstanceIn(ActivityScope)` provider, `provideActivityBrowserMode` in `TabRepositoryActivityModule`, which captures `browserModeStateHolder.currentMode.value` **once** at activity-component creation. The injected `currentBrowserMode` and the unqualified `TabRepository` (`provideUnqualifiedTabRepository`) both derive from that one frozen value, so they are always in agreement for the activity's lifetime.

So when `switchModeThen` runs an action immediately because `targetMode == currentBrowserMode`, it writes to the repository matching this instance's frozen mode — never the "wrong" one. The app-scoped holder flipping to the new mode does not affect this instance's repo; the pending switch simply recreates the activity afterward, and the next instance captures the new mode cleanly. This is documented in both `TabRepositoryActivityModule`'s KDoc and `switchModeThen`'s ("self-correcting … it simply re-stashes"). No change needed.

File: app/src/main/java/com/duckduckgo/app/browser/PendingModeSwitch.kt [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            bundle.putString(KEY_QUERY, pendingAction.query)
+            bundle.putString(KEY_SOURCE_TAB_ID, pendingAction.sourceTabId)
+            bundle.putBoolean(KEY_SKIP_HOME, pendingAction.skipHome)
+            bundle.putBoolean(KEY_IS_EXTERNAL, pendingAction.isExternal)
+        }
+    }
+    return bundle
+}
+
+internal fun Bundle.toPendingModeSwitch(): PendingModeSwitch? {
```
Comment (malmstein) [inline app/src/main/java/com/duckduckgo/app/browser/PendingModeSwitch.kt:68]: `toBundle` / `toPendingModeSwitch` is the load-bearing part here — if it doesn't round-trip cleanly the deferred action is silently lost across the recreate. it's pure and Android-testable (same `AndroidJUnit4` setup as `RealTabTitleResolverTest` from #8714) — can we add a round-trip test covering both `PendingAction` variants plus the null/garbage-bundle paths (unknown action, bad mode name, missing intent) before this merges?

**Review by malmstein — APPROVED** [review]
LGTM, approving. did not install or smoke-test the APK — static-only. nice refactor: generalising the old `pendingFireToRegularIntent` into a typed `PendingModeSwitch` (`ProcessIntent` / `OpenNewTab`) carried across the recreate is clean, and the `switchModeThen` design that re-enters itself on the recreated instance is genuinely well thought through — the doc comments and the explicit intent-redirection security invariant on `runAction` are exactly what I'd want to see on this path. the rejection fallback only reaches `OpenNewTab(FIRE)` when fire is unavailable, which is the right behaviour.

one non-blocking ask before merge: please add the round-trip test for `toBundle`/`toPendingModeSwitch` — it's the piece most likely to break quietly and the easiest thing here to cover. and since I didn't run it, worth a quick device check of the chat-menu "new tab" / "new fire tab" entries and an external `LAUNCH_REQUIRES_REGULAR_MODE` intent arriving while in fire mode, to confirm the deferred action survives the recreate.

File: app/src/main/java/com/duckduckgo/app/browser/PendingModeSwitch.kt (hunk 3 above)
Comment (0nko) [reply app/src/main/java/com/duckduckgo/app/browser/PendingModeSwitch.kt:68]: Added `PendingModeSwitchTest` (`AndroidJUnit4`, same setup as `RealTabTitleResolverTest`) — round-trips both `PendingAction` variants plus the null paths: empty bundle, unknown mode name, missing action, unknown action, and `ProcessIntent` with a missing intent. The malformed-bundle cases start from a valid bundle and corrupt one field, so a stale wire-format key fails the test rather than silently passing. `12f228066c`.
